File: core/src/com/megaman/game/screens/menus/impl/main/MainScreen.java
package com.megaman.game.screens.menus.impl.main;

import com.badlogic.gdx.Gdx;
import com.badlogic.gdx.graphics.g2d.Sprite;
import com.badlogic.gdx.graphics.g2d.TextureAtlas;
import com.badlogic.gdx.graphics.g2d.TextureRegion;
import com.badlogic.gdx.math.Vector2;
import com.badlogic.gdx.math.Vector3;
import com.badlogic.gdx.utils.Array;
import com.badlogic.gdx.utils.ObjectMap;
import com.megaman.game.MegamanGame;
import com.megaman.game.ViewVals;
import com.megaman.game.assets.MusicAsset;
import com.megaman.game.assets.SoundAsset;
import com.megaman.game.assets.TextureAsset;
import com.megaman.game.screens.ScreenEnum;
import com.megaman.game.screens.menus.MenuButton;
import com.megaman.game.screens.menus.MenuScreen;
import com.megaman.game.screens.menus.utils.BlinkingArrow;
import com.megaman.game.screens.menus.utils.ScreenSlide;
import com.megaman.game.screens.utils.TextHandle;
import com.megaman.game.sprites.SpriteDrawer;
import com.megaman.game.utils.ConstFuncs;
import com.megaman.game.utils.enums.Direction;
import com.megaman.game.utils.objs.Timer;
import com.megaman.game.world.WorldVals;
import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;

public class MainScreen extends MenuScreen {

    @RequiredArgsConstructor(access = AccessLevel.PACKAGE)
    public enum MainBtn {

        GAME_START("GAME START"),
        PASS_WORD("PASS WORD"),
        SETTINGS("SETTINGS"),
        CREDITS("CREDITS"),
        EXTRAS("EXTRAS"),
        EXIT("EXIT");

        public final String prompt;

    }

    @RequiredArgsConstructor(access = AccessLevel.PACKAGE)
    public enum SettingsBtn {

        BACK("BACK"),
        MUSIC_VOLUME("MUSIC: "),
        SOUND_EFFECTS_VOLUME("SOUND: ");

        public final String prompt;

    }

    private static final float SETTINGS_ARROW_BLINK_DUR = .3f;
    private static final float SETTINGS_TRANS_DUR = .5f;
    private static final Vector3 SETTINGS_TRAJ = new Vector3(15f * WorldVals.PPM, 0f, 0f);

    private final Sprite pose;
    private final Sprite title;
    private final Sprite subtitle;
    private final ScreenSlide screenSlide;
    private final Array<TextHandle> fonts;
    private final Array<Sprite> settingsArrs;
    private final Timer settingsArrBlinkTimer;
    private final ObjectMap<String, BlinkingArrow> blinkArrs;

    private boolean settingsArrowBlink;

    public MainScreen(MegamanGame game) {
        super(game, MainBtn.GAME_START.name());
        pose = new Sprite();
        title = new Sprite();
        subtitle = new Sprite();
        fonts = new Array<>();
        settingsArrs = new Array<>();
        blinkArrs = new ObjectMap<>();
        screenSlide = new ScreenSlide(uiCam, SETTINGS_TRAJ, ConstFuncs.getCamInitPos(),
                ConstFuncs.getCamInitPos().add(SETTINGS_TRAJ), SETTINGS_TRANS_DUR, true);
        settingsArrBlinkTimer = new Timer(SETTINGS_ARROW_BLINK_DUR);
        float row = .175f * WorldVals.PPM;
        for (MainBtn mainBtn : MainBtn.values()) {
            fonts.add(new TextHandle(new Vector2(2f * WorldVals.PPM, row * WorldVals.PPM), mainBtn.prompt));
            Vector2 arrCenter = new Vector2(1.5f * WorldVals.PPM, (row - (.0075f * WorldVals.PPM)) * WorldVals.PPM);
            blinkArrs.put(mainBtn.name(), new BlinkingArrow(assMan, arrCenter));
            row -= WorldVals.PPM * .025f;
        }
        row = .4f * WorldVals.PPM;
        for (SettingsBtn sBtn : SettingsBtn.values()) {
            fonts.add(new TextHandle(new Vector2(17f * WorldVals.PPM, row * WorldVals.PPM), sBtn.prompt));
            Vector2 arrCenter = new Vector2(16.5f * WorldVals.PPM, (row - (.0075f * WorldVals.PPM)) * WorldVals.PPM);
            blinkArrs.put(sBtn.name(), new BlinkingArrow(assMan, arrCenter));
            row -= WorldVals.PPM * .025f;
        }
        fonts.add(new TextHandle(new Vector2(.15f * WorldVals.PPM, .5f * WorldVals.PPM), "© OLD LAVY GENES, 20XX"));
        fonts.add(new TextHandle(new Vector2(21f * WorldVals.PPM, 12f * WorldVals.PPM),
                () -> "" + game.getAudioMan().getMusicVolume()));
        fonts.add(new TextHandle(new Vector2(21f * WorldVals.PPM, 11.2f * WorldVals.PPM),
                () -> "" + game.getAudioMan().getSoundVolume()));
        TextureRegion arrowReg = game.getAssMan().getTextureRegion(TextureAsset.UI_1, "Arrow");
        float y = 11.55f;
        for (int i = 0; i < 4; i++) {
            if (i != 0 && i % 2 == 0) {
                y -= .85f;
            }
            Sprite blinkArr = new Sprite(arrowReg);
            blinkArr.setBounds(
                    (i % 2 == 0 ? 20.25f : 22.5f) * WorldVals.PPM,
                    y * WorldVals.PPM,
                    WorldVals.PPM / 2f,
                    WorldVals.PPM / 2f);
            blinkArr.setFlip(i % 2 == 0, false);
            settingsArrs.add(blinkArr);
        }
        TextureAtlas atlas = game.getAssMan().getTextureAtlas(TextureAsset.UI_1);
        title.setRegion(atlas.findRegion("MegamanTitle"));
        title.setBounds(WorldVals.PPM, 8.25f * WorldVals.PPM, 13.25f * WorldVals.PPM, 5f * WorldVals.PPM);
        subtitle.setRegion(atlas.findRegion("Subtitle8bit"));
        subtitle.setSize(8f * WorldVals.PPM, 8f * WorldVals.PPM);
        subtitle.setCenter(ViewVals.VIEW_WIDTH * WorldVals.PPM / 2f, (ViewVals.VIEW_HEIGHT + 1) * WorldVals.PPM / 2f);
        pose.setRegion(atlas.findRegion("MegamanMaverick"));
        pose.setBounds(8f * WorldVals.PPM, 0f, 8f * WorldVals.PPM, 8f * WorldVals.PPM);
    }

    @Override
    public void show() {
        super.show();
        uiCam.position.set(ConstFuncs.getCamInitPos());
        game.getAudioMan().playMusic(MusicAsset.MM11_WILY_STAGE_MUSIC, true);
    }

    @Override
    public void render(float delta) {
        super.render(delta);
        if (!game.isPaused()) {
            screenSlide.update(delta);
            if (screenSlide.isJustFinished()) {
                screenSlide.reverse();
            }
            blinkArrs.get(getCurrBtnKey()).update(delta);
            settingsArrBlinkTimer.update(delta);
            if (settingsArrBlinkTimer.isFinished()) {
                settingsArrowBlink = !settingsArrowBlink;
                settingsArrBlinkTimer.reset();
            }
        }
        batch.setProjectionMatrix(uiCam.combined);
        batch.begin();
        blinkArrs.get(getCurrBtnKey()).draw(batch);
        SpriteDrawer.draw(title, batch);
        SpriteDrawer.draw(pose, batch);
        SpriteDrawer.draw(subtitle, batch);
        for (TextHandle font : fonts) {
            font.draw(batch);
        }
        if (settingsArrowBlink) {
            for (Sprite a : settingsArrs) {
                SpriteDrawer.draw(a, batch);
            }
        }
        batch.end();
    }

    @Override
    protected void onAnyMovement() {
        audioMan.playMusic(SoundAsset.CURSOR_MOVE_BLOOP_SOUND);
    }

    @Override
    protected boolean onAnySelection() {
        boolean allow = screenSlide.isFinished();
        if (allow) {
            audioMan.playMusic(SoundAsset.SELECT_PING_SOUND);
        }
        return allow;
    }

    @Override
    protected ObjectMap<String, MenuButton> defineMenuButtons() {
        return new ObjectMap<>() {{
            put(MainBtn.GAME_START.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    game.setScreen(game.getScreen(ScreenEnum.BOSS_SELECT));
                    return true;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case UP -> setMenuButton(MainBtn.EXIT.name());
                        case DOWN -> setMenuButton(MainBtn.PASS_WORD.name());
                    }
                }
            });
            put(MainBtn.PASS_WORD.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    // TODO: Set to password screen
                    return false;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case UP -> setMenuButton(MainBtn.GAME_START.name());
                        case DOWN -> setMenuButton(MainBtn.SETTINGS.name());
                    }
                }
            });
            put(MainBtn.SETTINGS.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    // TODO: Set to settings screen
                    screenSlide.init();
                    setMenuButton(SettingsBtn.BACK.name());
                    return false;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case UP -> setMenuButton(MainBtn.PASS_WORD.name());
                        case DOWN -> setMenuButton(MainBtn.CREDITS.name());
                    }
                }
            });
            put(MainBtn.CREDITS.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    return false;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case UP -> setMenuButton(MainBtn.SETTINGS.name());
                        case DOWN -> setMenuButton(MainBtn.EXTRAS.name());
                    }
                }
            });
            put(MainBtn.EXTRAS.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    game.setScreen(game.getScreen(ScreenEnum.EXTRAS));
                    return true;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case UP -> setMenuButton(MainBtn.CREDITS.name());
                        case DOWN -> setMenuButton(MainBtn.EXIT.name());
                    }
                }
            });
            put(MainBtn.EXIT.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    // TODO: Pop up dialog asking to confirm exit game, press X to accept, any other to abort
                    Gdx.app.exit();
                    return true;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case UP -> setMenuButton(MainBtn.EXTRAS.name());
                        case DOWN -> setMenuButton(MainBtn.GAME_START.name());
                    }
                }
            });
            put(SettingsBtn.BACK.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    screenSlide.init();
                    setMenuButton(MainBtn.SETTINGS.name());
                    return false;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case UP -> setMenuButton(SettingsBtn.SOUND_EFFECTS_VOLUME.name());
                        case DOWN -> setMenuButton(SettingsBtn.MUSIC_VOLUME.name());
                    }
                }
            });
            put(SettingsBtn.MUSIC_VOLUME.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    return false;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case LEFT -> {
                            int volume = audioMan.getMusicVolume();
                            volume = volume == 0 ? 10 : volume - 1;
                            audioMan.setMusicVolume(volume);
                        }
                        case RIGHT -> {
                            int volume = audioMan.getMusicVolume();
                            volume = volume == 10 ? 0 : volume + 1;
                            audioMan.setMusicVolume(volume);
                        }
                        case UP -> setMenuButton(SettingsBtn.BACK.name());
                        case DOWN -> setMenuButton(SettingsBtn.SOUND_EFFECTS_VOLUME.name());
                    }
                }
            });
            put(SettingsBtn.SOUND_EFFECTS_VOLUME.name(), new MenuButton() {
                @Override
                public boolean onSelect(float delta) {
                    return false;
                }

                @Override
                public void onNavigate(Direction direction, float delta) {
                    switch (direction) {
                        case LEFT -> {
                            int volume = audioMan.getSoundVolume();
                            volume = volume == 0 ? 10 : volume - 1;
                            audioMan.setSoundVolume(volume);
                        }
                        case RIGHT -> {
                            int volume = audioMan.getSoundVolume();
                            volume = volume == 10 ? 0 : volume + 1;
                            audioMan.setSoundVolume(volume);
                        }
                        case UP -> setMenuButton(SettingsBtn.MUSIC_VOLUME.name());
                        case DOWN -> setMenuButton(SettingsBtn.BACK.name());
                    }
                }
            });
        }};
    }

}

File: core/src/com/megaman/maverick/game/screens/menus/MainScreen.kt
package com.megaman.maverick.game.screens.menus

import com.badlogic.gdx.Gdx
import com.badlogic.gdx.graphics.g2d.Sprite
import com.badlogic.gdx.math.Vector2
import com.badlogic.gdx.math.Vector3
import com.badlogic.gdx.utils.Array
import com.badlogic.gdx.utils.ObjectMap
import com.engine.common.GameLogger
import com.engine.common.enums.Direction
import com.engine.common.extensions.getTextureAtlas
import com.engine.common.extensions.getTextureRegion
import com.engine.common.extensions.objectMapOf
import com.engine.common.time.Timer
import com.engine.drawables.fonts.BitmapFontHandle
import com.engine.screens.menus.IMenuButton
import com.megaman.maverick.game.ConstVals
import com.megaman.maverick.game.MegamanMaverickGame
import com.megaman.maverick.game.assets.MusicAsset
import com.megaman.maverick.game.assets.SoundAsset
import com.megaman.maverick.game.assets.TextureAsset
import com.megaman.maverick.game.screens.ScreenEnum
import com.megaman.maverick.game.screens.utils.BlinkingArrow
import com.megaman.maverick.game.screens.utils.ScreenSlide
import com.megaman.maverick.game.utils.MegaUtilMethods.getDefaultFontSize
import com.megaman.maverick.game.utils.getDefaultCameraPosition
import com.megaman.maverick.game.utils.setToDefaultPosition

class MainScreen(game: MegamanMaverickGame) :
    AbstractMenuScreen(game, MainScreenButton.GAME_START.text) {

  enum class MainScreenButton(val text: String) {
    GAME_START("GAME START"),
    PASSWORD("PASSWORD"),
    SETTINGS("SETTINGS"),
    CREDITS("CREDITS"),
    EXTRAS("EXTRA"),
    EXIT("EXIT")
  }

  enum class MainScreenSettingsButton(val text: String) {
    BACK("BACK"),
    MUSIC_VOLUME("MUSIC VOLUME"),
    SOUND_EFFECTS_VOLUME("SOUND EFFECTS VOLUME"),
  }

  companion object {
    const val TAG = "MainScreen"
    private const val SETTINGS_ARROW_BLINK_DUR = .3f
    private const val SETTINGS_TRANS_DUR = .5f
    private val SETTINGS_TRAJ = Vector3(15f * ConstVals.PPM, 0f, 0f)
  }

  override val menuButtons = objectMapOf<String, IMenuButton>()

  private val pose = Sprite()
  private val title = Sprite()
  private val subtitle = Sprite()

  private var screenSlide =
      ScreenSlide(
          castGame.getUiCamera(),
          SETTINGS_TRAJ,
          getDefaultCameraPosition(),
          getDefaultCameraPosition().add(SETTINGS_TRAJ),
          SETTINGS_TRANS_DUR,
          true)

  private val fontHandles = Array<BitmapFontHandle>()
  private val settingsArrows = Array<Sprite>()

  private val settingsArrowBlinkTimer = Timer(SETTINGS_ARROW_BLINK_DUR)
  private val blinkArrows = ObjectMap<String, BlinkingArrow>()

  private var settingsArrowBlink = false

  init {
    var row = 0.175f * ConstVals.PPM

    MainScreenButton.values().forEach {
      val fontHandle =
          BitmapFontHandle(
              { it.text },
              getDefaultFontSize(),
              Vector2(2f * ConstVals.PPM, row * ConstVals.PPM),
              centerX = false,
              centerY = false,
              fontSource = "Megaman10Font.ttf")
      fontHandles.add(fontHandle)
      val arrowCenter =
          Vector2(1.5f * ConstVals.PPM, (row - (.0075f * ConstVals.PPM)) * ConstVals.PPM)
      blinkArrows.put(it.text, BlinkingArrow(game.assMan, arrowCenter))
      row -= ConstVals.PPM * .025f
    }

    row = 0.4f * ConstVals.PPM

    MainScreenSettingsButton.values().forEach {
      val fontHandle =
          BitmapFontHandle(
              { it.text },
              getDefaultFontSize(),
              Vector2(17f * ConstVals.PPM, row * ConstVals.PPM),
              centerX = false,
              centerY = false,
              fontSource = "Megaman10Font.ttf")
      fontHandles.add(fontHandle)
      val arrowCenter =
          Vector2(16.5f * ConstVals.PPM, (row - (.0075f * ConstVals.PPM)) * ConstVals.PPM)
      blinkArrows.put(it.text, BlinkingArrow(game.assMan, arrowCenter))
    }

    fontHandles.add(
        BitmapFontHandle(
            { "© OLD LAVY GENES, 20XX" },
            getDefaultFontSize(),
            Vector2(0.15f * ConstVals.PPM, 0.5f * ConstVals.PPM),
            centerX = false,
            centerY = false,
            fontSource = "Megaman10Font.ttf"))

    fontHandles.add(
        BitmapFontHandle(
            { castGame.audioMan.musicVolume.toString() },
            getDefaultFontSize(),
            Vector2(21f * ConstVals.PPM, 12f * ConstVals.PPM),
            centerX = false,
            centerY = false,
            fontSource = "Megaman10Font.ttf"))

    fontHandles.add(
        BitmapFontHandle(
            { castGame.audioMan.soundVolume.toString() },
            getDefaultFontSize(),
            Vector2(21f * ConstVals.PPM, 11.2f * ConstVals.PPM),
            centerX = false,
            centerY = false,
            fontSource = "Megaman10Font.ttf"))

    val arrowRegion = game.assMan.getTextureRegion(TextureAsset.UI_1.source, "Arrow")
    var y = 11.55f
    for (i in 0 until 4) {
      if (i != 0 && i % 2 == 0) y -= 0.85f
      val blinkArrow = Sprite(arrowRegion)
      blinkArrow.setBounds(
          (if (i % 2 == 0) 20.25f else 22.5f) * ConstVals.PPM,
          y * ConstVals.PPM,
          ConstVals.PPM / 2f,
          ConstVals.PPM / 2f)
      blinkArrow.setFlip(i % 2 == 0, false)
      settingsArrows.add(blinkArrow)
    }

    val atlas = game.assMan.getTextureAtlas(TextureAsset.UI_1.source)
    title.setRegion(atlas.findRegion("MegamanTitle"))
    title.setBounds(
        ConstVals.PPM.toFloat(), 8.25f * ConstVals.PPM, 13.25f * ConstVals.PPM, 5f * ConstVals.PPM)
    subtitle.setRegion(atlas.findRegion("Subtitle8bit"))
    subtitle.setSize(8f * ConstVals.PPM, 8f * ConstVals.PPM)
    subtitle.setCenter(
        ConstVals.VIEW_WIDTH * ConstVals.PPM / 2f, (ConstVals.VIEW_HEIGHT + 1) * ConstVals.PPM / 2f)
    pose.setRegion(atlas.findRegion("MegamanMaverick"))
    pose.setBounds(8f * ConstVals.PPM, 0f, 8f * ConstVals.PPM, 8f * ConstVals.PPM)

    menuButtons.put(
        MainScreenButton.GAME_START.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            castGame.setCurrentScreen(ScreenEnum.BOSS_SELECT.name)
            return true
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.UP -> MainScreenButton.EXIT.text
              Direction.DOWN -> MainScreenButton.PASSWORD.text
              else -> null
            }
          }
        })

    menuButtons.put(
        MainScreenButton.PASSWORD.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            // TODO: game.setCurrentScreen(ScreenEnum.PASSWORD_SCREEN.name)
            return false
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.UP -> MainScreenButton.GAME_START.text
              Direction.DOWN -> MainScreenButton.SETTINGS.text
              else -> null
            }
          }
        })

    menuButtons.put(
        MainScreenButton.SETTINGS.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            screenSlide.init()
            currentButtonKey = MainScreenSettingsButton.BACK.text
            return false
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.UP -> MainScreenButton.PASSWORD.text
              Direction.DOWN -> MainScreenButton.CREDITS.text
              else -> null
            }
          }
        })

    menuButtons.put(
        MainScreenButton.CREDITS.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            return false
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.UP -> MainScreenButton.SETTINGS.text
              Direction.DOWN -> MainScreenButton.EXTRAS.text
              else -> null
            }
          }
        })

    menuButtons.put(
        MainScreenButton.EXTRAS.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            game.setCurrentScreen(ScreenEnum.EXTRAS.name)
            return false
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.UP -> MainScreenButton.CREDITS.text
              Direction.DOWN -> MainScreenButton.EXIT.text
              else -> null
            }
          }
        })

    menuButtons.put(
        MainScreenButton.EXIT.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            Gdx.app.exit()
            return true
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.UP -> MainScreenButton.EXTRAS.text
              Direction.DOWN -> MainScreenButton.GAME_START.text
              else -> null
            }
          }
        })

    menuButtons.put(
        MainScreenSettingsButton.BACK.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            screenSlide.init()
            currentButtonKey = MainScreenButton.SETTINGS.text
            return false
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.UP -> MainScreenSettingsButton.SOUND_EFFECTS_VOLUME.text
              Direction.DOWN -> MainScreenSettingsButton.MUSIC_VOLUME.text
              else -> null
            }
          }
        })

    menuButtons.put(
        MainScreenSettingsButton.MUSIC_VOLUME.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            return false
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.LEFT -> {
                var volume = castGame.audioMan.musicVolume
                volume = if (volume == 0f) 10f else volume - 1f
                castGame.audioMan.musicVolume = volume
                null
              }
              Direction.RIGHT -> {
                var volume = castGame.audioMan.musicVolume
                volume = if (volume == 10f) 0f else volume + 1f
                castGame.audioMan.musicVolume = volume
                null
              }
              Direction.UP -> MainScreenSettingsButton.BACK.text
              Direction.DOWN -> MainScreenSettingsButton.SOUND_EFFECTS_VOLUME.text
            }
          }
        })

    menuButtons.put(
        MainScreenSettingsButton.SOUND_EFFECTS_VOLUME.text,
        object : IMenuButton {
          override fun onSelect(delta: Float): Boolean {
            return false
          }

          override fun onNavigate(direction: Direction, delta: Float): String? {
            return when (direction) {
              Direction.LEFT -> {
                var volume = castGame.audioMan.soundVolume
                volume = if (volume == 0f) 10f else volume - 1f
                castGame.audioMan.soundVolume = volume
                null
              }
              Direction.RIGHT -> {
                var volume = castGame.audioMan.soundVolume
                volume = if (volume == 10f) 0f else volume + 1f
                castGame.audioMan.soundVolume = volume
                null
              }
              Direction.UP -> MainScreenSettingsButton.MUSIC_VOLUME.text
              Direction.DOWN -> MainScreenSettingsButton.BACK.text
            }
          }
        })
  }

  override fun show() {
    super.show()
    castGame.getUiCamera().setToDefaultPosition()
    castGame.audioMan.playMusic(MusicAsset.MM_OMEGA_TITLE_THEME_MUSIC)
  }

  override fun render(delta: Float) {
    super.render(delta)
    if (!game.paused) {
      screenSlide.update(delta)
      if (screenSlide.justFinished) screenSlide.reverse()

      blinkArrows.get(currentButtonKey).update(delta)

      settingsArrowBlinkTimer.update(delta)
      if (settingsArrowBlinkTimer.isFinished()) {
        settingsArrowBlink = !settingsArrowBlink
        settingsArrowBlinkTimer.reset()
      }
    }

    val batch = game.batch
    batch.projectionMatrix = castGame.getUiCamera().combined
    batch.begin()

    blinkArrows.get(currentButtonKey).draw(batch)
    title.draw(batch)
    pose.draw(batch)
    subtitle.draw(batch)

    fontHandles.forEach { it.draw(batch) }

    if (settingsArrowBlink) settingsArrows.forEach { it.draw(batch) }

    batch.end()
  }

  override fun onAnyMovement() {
    super.onAnyMovement()
    GameLogger.debug(TAG, "Current button: $currentButtonKey")
    castGame.audioMan.playSound(SoundAsset.CURSOR_MOVE_BLOOP_SOUND)
  }

  override fun onAnySelection(): Boolean {
    val allow = screenSlide.finished
    if (allow) castGame.audioMan.playSound(SoundAsset.SELECT_PING_SOUND)
    return allow
  }
}
